### ISSUE TITLE
AMDGPU backend crashes on atomicrmw instruction when using -disable-promote-alloca-to-lds

### ISSUE THREAD

**Issue by ivanrodriguez3753** [llvm/llvm-project#54274]
```
source_filename = "llvm-link"
target datalayout = "e-p:64:64-p1:64:64-p2:32:32-p3:32:32-p4:64:64-p5:32:32-p6:32:32-i64:64-v16:16-v24:32-v32:32-v48:64-v96:128-v192:256-v256:256-v512:512-v1024:1024-v2048:2048-n32:64-S32-A5-G1-ni:7"
target triple = "amdgcn-amd-amdhsa"

; Function Attrs: alwaysinline convergent norecurse nounwind
define weak amdgpu_kernel void @__omp_offloading_67_6b1f7ec1_main_l3(i64 %A) #0 {
entry:
  %A.addr = alloca i64, align 8, addrspace(5)
  %A.addr.ascast = addrspacecast i64 addrspace(5)* %A.addr to i64*
  store i64 %A, i64* %A.addr.ascast, align 8
  %conv = bitcast i64* %A.addr.ascast to i32*
  %0 = atomicrmw add i32* %conv, i32 1 monotonic, align 4
  ret void
}

attributes #0 = { alwaysinline convergent norecurse nounwind "amdgpu-flat-work-group-size"="1,1024" "amdgpu-implicitarg-num-bytes"="56" "frame-pointer"="none" "min-legal-vector-width"="0" "no-trapping-math"="true" "omp_target_num_teams"="1" "stack-protector-buffer-size"="8" "target-cpu"="gfx908" "target-features"="+16-bit-insts,+ci-insts,+dl-insts,+dot1-insts,+dot2-insts,+dot3-insts,+dot4-insts,+dot5-insts,+dot6-insts,+dot7-insts,+dpp,+flat-address-space,+gfx8-insts,+gfx9-insts,+mai-insts,+s-memrealtime,+s-memtime-inst" "uniform-work-group-size"="true" }
```
The LLVM IR above is crashing and produces the stack trace:
```
LLVM ERROR: Cannot select: t16: i32,ch = AtomicLoadAdd<(load store syncscope("agent") monotonic (s32) on %ir."$$_a_t5", addrspace 5)> t18, FrameIndex:i32<0>, Constant:i32<1>
  t11: i32 = FrameIndex<0>
  t15: i32 = Constant<1>
In function: targetatomic_simple_$ck_L5_1
PLEASE submit a bug report to Cray and include the crash backtrace.
Stack dump:
0.      Program arguments: llc -mtriple=amdgcn-amd-amdhsa -disable-promote-alloca-to-lds -mcpu=gfx906 -filetype=asm a.out.bc 
1.      Running pass 'CallGraph Pass Manager' on module 'a.out.bc'.
2.      Running pass 'AMDGPU DAG->DAG Pattern Instruction Selection' on function '@"targetatomic_simple_$ck_L5_1"'
 #0 0x0000000002efd62e PrintStackTraceSignalHandler(void*) Signals.cpp:0:0
 #1 0x0000000002efa99d SignalHandler(int) Signals.cpp:0:0
 #2 0x00007f65969888bf (/lib64/libpthread.so.0+0x168bf)
 #3 0x00007f65956b8cba raise (/lib64/libc.so.6+0x4acba)
 #4 0x00007f65956ba354 abort (/lib64/libc.so.6+0x4c354)
 #5 0x0000000002e4f9d2 llvm::report_fatal_error(llvm::Twine const&, bool) 
 #6 0x0000000002cfc0dd llvm::SelectionDAGISel::CannotYetSelect(llvm::SDNode*)
 #7 0x0000000002d0803a llvm::SelectionDAGISel::SelectCodeCommon(llvm::SDNode*, unsigned char const*, unsigned int)  
 #8 0x0000000000ee4d40 AMDGPUDAGToDAGISel::Select(llvm::SDNode*)
 #9 0x0000000002cfde88 llvm::SelectionDAGISel::DoInstructionSelection() 
#10 0x0000000002d00dc9 llvm::SelectionDAGISel::CodeGenAndEmitDAG() 
#11 0x0000000002d03bcd llvm::SelectionDAGISel::SelectAllBasicBlocks(llvm::Function const&) 
#12 0x0000000002d0572a llvm::SelectionDAGISel::runOnMachineFunction(llvm::MachineFunction&) 
#13 0x0000000000ee8599 AMDGPUDAGToDAGISel::runOnMachineFunction(llvm::MachineFunction&) 
#14 0x0000000002264ee5 llvm::MachineFunctionPass::runOnFunction(llvm::Function&) 
#15 0x00000000026e5c9e llvm::FPPassManager::runOnFunction(llvm::Function&) 
#16 0x0000000001e1a615 (anonymous namespace)::CGPassManager::runOnModule(llvm::Module&) 
#17 0x00000000026e6617 llvm::legacy::PassManagerImpl::run(llvm::Module&) 
#18 0x0000000000951bf1 compileModule(char**, llvm::LLVMContext&) 
#19 0x000000000088eaf1 main 
#20 0x00007f65956a329c __libc_start_main (/lib64/libc.so.6+0x3529c)
#21 0x00000000009482b9 _start 
```
Any pointers or explanation on the issue would be great. Thanks.

**Comment by ivanrodriguez3753** [comment]
-mcpu=gfx906
I'm an engineer at Cray

**Comment by jayfoad** [comment]
`test/CodeGen/AMDGPU/private-memory-atomics.ll` does something similar. To quote the comments in that file, "Private atomics have no real use" and "This works because promote allocas pass replaces these with LDS atomics". So after AMDGPUPromoteAlloca the atomicrmw works on a global in addrspace 3 instead of an alloca in addrspace 5.

**Comment by arsenm** [comment]
This is an interesting edge case. We could either stop optimizing flat atomics to private, or we need to do something to support private atomics in codegen.

For the MUBUF private path, we can directly encode most of the same atomic operations as global. For scratch, the instructions seem to not exist. Based on how FLAT encodings work I might guess the scratch atomics secretly exist, but we probably shouldn't rely on that.

Can we legally transform private atomics into the equivalent non-atomic operations, or do we have to codegen by casting back to flat?

**Comment by arsenm** [comment]
Also relates issue #54635

**Comment by ivanrodriguez3753** [comment]
@arsenm I would argue this is a bug, not an edge case. If AMD is relying on users NOT using a certain optimization (in this case, relying on the user NOT passing the `disable-promote-alloca-to-lds`) they should either force the optimization to happen (remove the option of using `disable-promote-alloca-to-lds`) or implement a pass that does what you said (transform private atomic to its equivalent non-atomic operation) on compilations that included the `disable-promote-alloca-to-lds` flag. Thoughts?

**Comment by arsenm** [comment]
It's a bug hit on an edge case, which an optimization helped hide. Real code using atomics can't be optimized to something that doesn't require atomics, which is what this toy example is.

**Comment by ivanrodriguez3753** [comment]
The comment in `test/CodeGen/AMDGPU/private-memory-atomics.ll` makes it seem like this was intentionally left unimplemented, as opposed to the bug accidentally being hidden by the optimization. The test explicitly states that it relies on that optimization, but the flag is still available upstream so that's a little confusing (amdclang gives some sort of unused flag warning when attempting to pass `-disable-promote-alloca-to-lds`).

I would also argue that real code using atomics CAN be optimized to something that doesn't require atomics. In fact, the 'sanity check' C test case that triggered this was
```
int main() {
    int A = 1;
    #pragma omp target
    #pragma omp atomic
        A += 1;
}
```
I believe this code is legal, although redundant. Redundancy shouldn't cause a compiler abort, though (a warning seems reasonable). Since the use of atomics is redundant, it can be optimized to something that doesn't require atomics. What constitutes "real code"?
On top of AMD's LDS memory use being non-transparent to the user, its use is also inconsistent. The amdgcn assembly generated by `amdclang` has a comment `LDSByteSize` which does not scale in an intuitive way, so presumably there are checks on whether or not to use LDS based on many factors, such as the number of variables that could possibly be promoted to LDS. For example, this test case (although contrived as written, could very realistically be used as part of a larger test case through inlining/other optimizations that replace certain variables with private variables) breaks amdclang with the following invocation:
`/opt/rocm-5.0.2/bin/amdclang -fopenmp -O2 --offload-arch=gfx906 tooManyLocals.c`
```
int main() {
    int myA1 = 123;
    int myA2 = 123;
    int myA3 = 123;
    int myA4 = 123;
    int myA5 = 123;
    int myA6 = 123;
    int myA7 = 123;
    int myA8 = 123;
    int myA9 = 123;

#pragma omp target
    {
#pragma omp atomic
        myA1++;
#pragma omp atomic
        myA2++;
#pragma omp atomic
        myA3++;
#pragma omp atomic
        myA4++;
#pragma omp atomic
        myA5++;
#pragma omp atomic
        myA6++;
#pragma omp atomic
        myA7++;
#pragma omp atomic
        myA8++;
#pragma omp atomic
        myA9++;
    }
}
```
```
rodriiva@pe31:/ptmp/rodriiva/hpc-pe-cce-cce/sandboxSymLink/CAST-27098/ldsProfiling/tooManyLocals> /opt/rocm-5.0.2/bin/amdclang -fopenmp -O2 --offload-arch=gfx906 tooManyLocals.c 
LLVM ERROR: Cannot select: 0x54e5a88: i32,ch = AtomicLoadAdd<(load store monotonic (s32) on %ir.115, addrspace 5)> 0x54e4ad8:1, FrameIndex:i32<6>, Constant:i32<1>
  0x546a438: i32 = FrameIndex<6>
  0x546b068: i32 = Constant<1>
In function: __omp_offloading_49_76b1cc7b_main_l13
PLEASE submit a bug report to https://bugs.llvm.org/ and include the crash backtrace.
Stack dump:
0.      Program arguments: /opt/rocm-5.0.2/llvm/bin/llc /tmp/tooManyLocals-be21a1-gfx906-optimized-d0a893.bc -O2 -mtriple=amdgcn-amd-amdhsa -mcpu=gfx906 -filetype=obj --amdhsa-code-object-version=4 -o /tmp/tooManyLocals-be21a1-gfx906-e5216c.o
1.      Running pass 'CallGraph Pass Manager' on module '/tmp/tooManyLocals-be21a1-gfx906-optimized-d0a893.bc'.
2.      Running pass 'AMDGPU DAG->DAG Pattern Instruction Selection' on function '@__omp_offloading_49_76b1cc7b_main_l13'
 #0 0x0000000001d4688f PrintStackTraceSignalHandler(void*) Signals.cpp:0:0
 #1 0x0000000001d44279 SignalHandler(int) Signals.cpp:0:0
 #2 0x00007fd6d92508c0 __restore_rt (/lib64/libpthread.so.0+0x168c0)
 #3 0x00007fd6d7cc6cdb raise (/lib64/libc.so.6+0x4acdb)
 #4 0x00007fd6d7cc8375 abort (/lib64/libc.so.6+0x4c375)
 #5 0x0000000001ca6025 llvm::report_fatal_error(llvm::Twine const&, bool) (/opt/rocm-5.0.2/llvm/bin/llc+0x1ca6025)
 #6 0x0000000001b83df2 llvm::SelectionDAGISel::CannotYetSelect(llvm::SDNode*) (/opt/rocm-5.0.2/llvm/bin/llc+0x1b83df2)
 #7 0x0000000001b86e4a llvm::SelectionDAGISel::SelectCodeCommon(llvm::SDNode*, unsigned char const*, unsigned int) (/opt/rocm-5.0.2/llvm/bin/llc+0x1b86e4a)
 #8 0x000000000097c2e5 AMDGPUDAGToDAGISel::Select(llvm::SDNode*) (/opt/rocm-5.0.2/llvm/bin/llc+0x97c2e5)
 #9 0x0000000001b82a34 llvm::SelectionDAGISel::DoInstructionSelection() (/opt/rocm-5.0.2/llvm/bin/llc+0x1b82a34)
#10 0x0000000001b8a416 llvm::SelectionDAGISel::CodeGenAndEmitDAG() (/opt/rocm-5.0.2/llvm/bin/llc+0x1b8a416)
#11 0x0000000001b8dc6c llvm::SelectionDAGISel::SelectAllBasicBlocks(llvm::Function const&) (/opt/rocm-5.0.2/llvm/bin/llc+0x1b8dc6c)
#12 0x0000000001b8fc12 llvm::SelectionDAGISel::runOnMachineFunction(llvm::MachineFunction&) (.part.884) SelectionDAGISel.cpp:0:0
#13 0x0000000000981b4a AMDGPUDAGToDAGISel::runOnMachineFunction(llvm::MachineFunction&) (/opt/rocm-5.0.2/llvm/bin/llc+0x981b4a)
#14 0x00000000011cf068 llvm::MachineFunctionPass::runOnFunction(llvm::Function&) (/opt/rocm-5.0.2/llvm/bin/llc+0x11cf068)
#15 0x00000000015f00b9 llvm::FPPassManager::runOnFunction(llvm::Function&) (/opt/rocm-5.0.2/llvm/bin/llc+0x15f00b9)
#16 0x0000000000de4bc9 (anonymous namespace)::CGPassManager::runOnModule(llvm::Module&) CallGraphSCCPass.cpp:0:0
#17 0x00000000015f1537 llvm::legacy::PassManagerImpl::run(llvm::Module&) (/opt/rocm-5.0.2/llvm/bin/llc+0x15f1537)
#18 0x00000000007484e1 compileModule(char**, llvm::LLVMContext&) llc.cpp:0:0
#19 0x00000000006c1d52 main (/opt/rocm-5.0.2/llvm/bin/llc+0x6c1d52)
#20 0x00007fd6d7cb12bd __libc_start_main (/lib64/libc.so.6+0x352bd)
#21 0x000000000074047a _start /home/abuild/rpmbuild/BUILD/glibc-2.26/csu/../sysdeps/x86_64/start.S:122:0
clang-14: error: unable to execute command: Aborted (core dumped)
clang-14: error: amdgcn-link command failed due to signal (use -v to see invocation)
AMD clang version 14.0.0 (https://github.com/RadeonOpenCompute/llvm-project roc-5.0.2 22065 030a405a181176f1a7749819092f4ef8ea5f0758)
Target: x86_64-unknown-linux-gnu
Thread model: posix
InstalledDir: /opt/rocm-5.0.2/llvm/bin
clang-14: note: diagnostic msg: Error generating preprocessed source(s).
```
The test case does not crash if you comment out all but one of the atomic updates, so I think the criteria for selecting which variables go to LDS is unintuitive at best.
Are you able to provide an ETA for this fix? We would like to request high priority as it is affecting us a fair amount (#54635 was filed by my teammate).

**Comment by arsenm** [comment]
> The comment in `test/CodeGen/AMDGPU/private-memory-atomics.ll` makes it seem like this was intentionally left unimplemented, as opposed to the bug accidentally being hidden by the optimization. 
Yes, essentially. This is just supposed to make sure this one pass pass doesn't blow up on it, not that private memory atomics were something really handled.


> The test explicitly states that it relies on that optimization, but the flag is still available upstream so that's a little confusing (amdclang gives some sort of unused flag warning when attempting to pass `-disable-promote-alloca-to-lds`).
It's not a user facing clang flag. It's a backend debug flag; from clang you need a -mllvm prefix to use it. This optimization is also a red herring, You could construct a testcase that would fail with or without this flag (as you did below)


> 
> I would also argue that real code using atomics CAN be optimized to something that doesn't require atomics. In fact, the 'sanity check' C test case that triggered this was
> 
> ```
> int main() {
>     int A = 1;
>     #pragma omp target
>     #pragma omp atomic
>         A += 1;
> }
> ```
> 
> I believe this code is legal, although redundant. Redundancy shouldn't cause a compiler abort, though (a warning seems reasonable). Since the use of atomics is redundant, it can be optimized to something that doesn't require atomics. What constitutes "real code"? 
It's legal, but also not doing any communication that would require atomics since a stack variable can't be accessed across lanes. It's just atomics for the sake of using atomics in a compiler test.


> On top of AMD's LDS memory use being non-transparent to the user, its use is also inconsistent. The amdgcn assembly generated by `amdclang` has a comment `LDSByteSize` which does not scale in an intuitive way, so presumably there are checks on whether or not to use LDS based on many factors, such as the number of variables that could possibly be promoted to LDS.
The backend will try to replace stack usage with LDS as long as it can fit within the function's occupancy budget, existing LDS usage, and required maximum group size. Since by default any function needs to support a 1024 workitem function, and we need maximum workgroup size * N bytes of LDS, that doesn't leave very many bytes available. This is part of why the attributes restricting the maximum group size can help performance.


> Are you able to provide an ETA for this fix? We would like to request high priority as it is affecting us a fair amount (#54635 was filed by my teammate).

I think this is probably 2-3 days of work to implement some kind of private atomic lowering scheme.

**Comment by arsenm** [comment]
Should be fixed by the stack of patches in https://reviews.llvm.org/D123171

**Comment by jtpham4** [comment]
On god, this is absolutely epic 😮‍💨🙏
-jtpham4

**Comment by arsenm** [comment]
Should be fixed by e6012c8e0fbdf422325e66c96537de03d258deb7
